### PR TITLE
[receiver/filelogreceiver] Truncate log , outputting the log unit per max_log_size

### PR DESCRIPTION
**Pull request by maokitty**
**Description:** <Describe what has changed.>

Truncate log if filelogreceiver scanner return error ErrTooLong 。There will be two condition for scanner return this error

1.  multiline regex patterns do not match the log,scanner will read until reach the max_log_size limit and throw ErrTooLong
2. multiline regex patterns work well but log entry is too large for a log unit

This commit will truncate the log unit to  max_log_size ,and emit it to next consumer

**Link to tracking Issue:** [#16487 ](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/16487)

**Testing:** <Describe what testing was performed and which tests were added.>
large log line test  which is large than max_log_size

### CODE REVIEW

**Review by djaglowski — COMMENTED** [review]
Thanks for the PR. This looks quite good. I have a few minor requests.

**Comment by djaglowski** [comment]
Interesting alternative in #17309. @maokitty, I'm curious to hear your assessment of the difference.

File: pkg/stanza/fileconsumer/reader_test.go [hunk 1]
```diff
@@ -93,18 +108,21 @@ func TestTokenization(t *testing.T) {
 
 			for _, expected := range tc.expected {
 				require.Equal(t, expected, readToken(t, emitChan))
+				if strings.Compare(tc.testName, "large_log_line") == 0 {
+					r.ReadToEnd(context.Background())
```
Comment (haoqixu) [inline pkg/stanza/fileconsumer/reader_test.go:112]: Can the unittest run `r.ReadToEnd()` twice?

**Comment by maokitty** [comment]
> Interesting alternative in #17309. @maokitty, I'm curious to hear your assessment of the difference.

First I am very happy that more people help to solve this problem ,Thank you for your PR  @haoqixu 
Here I found a case which i think [#17309 ](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/17309) can not solve it.

### Case
fileContent
```
2022-12-31 1 2022-12-31 2 
```
config
```
max_log_size: 15
multiline:
   line_start_pattern: \d+-\d+-\d+
```
expected output
```
2022-12-31 1 
2022-12-31 2 
```
output of [#17309 ](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/17309) 
```
2022-12-31 1 20
22-12-31 2 
```
### Reason
maxTokenSize ( max_log_size ) in scanner is used to guarantee no overflow , not for a log unit bytes limit threshold
what we need to handle is when "overflow" happens, we should find a way to make it through ,not loops forever

### My opinion 
 we should handle the case of ErrorTooLong to solve the problem of  loops forever , which is the  goal of this PR. 
then make another PR to handle "log unit bytes limit threshold".
people , including me , who is the first time to see config max_log_size will think it is use for limit log unit ,while is not after reading the code

### Suggest for read file desgin
use a fix size of buff read forever if there is more bytes , stop at EOF ,loop it ,handle log unit limit and split it in memory 

I think this may be is a "big project" , and it will be better to have another feature (  i am not sure about the process ) to talk about it

what do you think ?  @djaglowski    @haoqixu  @alexvanboxel 

### About PR comment
I think all the comment are good advice , like too long test string and read until there is no more to read other than wait for next loop will be better

after our discuss, then let's  make a decision of this PR


BTW: should we add more unit test of reader? all of the unit test now is using NewNewlineSplitFunc @djaglowski

**Comment by haoqixu** [comment]
> `2022-12-31 1 2022-12-31 2 `

When scanning this fileContent, this PR is actually using the buffer with the cap of `16384` bytes instead of `max_log_size` (it is `15` here) bytes. If you make the content line longger than `16384` bytes or use a buffer with cap of `max_log_size`, this patch will truncate the log entry at the same position as #17309

```go
// opentelemetry-collector-contrib/pkg/stanza/fileconsumer/scanner.go
// NewPositionalScanner creates a new positional scanner
func NewPositionalScanner(r io.Reader, maxLogSize int, startOffset int64, splitFunc bufio.SplitFunc) *PositionalScanner {
	ps := &PositionalScanner{
		pos:     startOffset,
		Scanner: bufio.NewScanner(r),
	}

	buf := make([]byte, 0, 16384)  // <- this line
	ps.Scanner.Buffer(buf, maxLogSize)

	scanFunc := func(data []byte, atEOF bool) (advance int, token []byte, err error) {
		advance, token, err = splitFunc(data, atEOF)
		ps.pos += int64(advance)
		return
	}
	ps.Scanner.Split(scanFunc)
	return ps
}
```

**Comment by haoqixu** [comment]
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/16487#issuecomment-1368021190

**Comment by maokitty** [comment]
> 

Yes, it will truncate log in the same position if one log unit is too large. and that's why scanner may need a "break change"

This PR actually is  a temp fix , and i think it is still meaningful ：

1.  people can choose an approprate max_log_size for their log, it is reasonable to limit log size
2. If one line is too large , with the  default newline config , otel just split it and can recombine it (this is the trick part),then still send it to backend
3.  If one line is too large, with line start/end config, the bad case occurs only in boundary ,we can still use recombine , but isn't it so weird for ask user to config both line start pattern and recombine is start entry?

Anyway, agent still work , and it's better than  loop forever.

**Comment by djaglowski** [comment]
Thanks again for the work on this @maokitty. I've merged #17358 as I believe it solves the looping problem as intended. Please let me know if you believe there is more to the problem that still requires work.

**Pull request closed without merging**